core/config.man
.SH CONFIG

The sake.yaml config is based on the following concepts:

.RS 2
.IP "\(bu" 2
\fBservers\fR are servers, local or remote, that have a host
.IP "\(bu" 2
\fBtasks\fR are shell commands that you write and then run for selected \fBservers\fR
.IP "\(bu" 2
\fBspecs\fR are configs that alter \fBtask\fR execution and output
.IP "\(bu" 2
\fBtargets\fR are configs that provide shorthand filtering of \fBservers\fR when executing tasks
.IP "\(bu" 2
\fBthemes\fR are used to modify the output of \fBsake\fR commands
.IP "" 0
.RE

\fBSpecs\fR, \fBtargets\fR and \fBthemes\fR come with a default setting that the user can override.

Check the files and environment section to see how the config file is loaded.

Below is a config file detailing all of the available options and their defaults.

.RS 4
 # Import servers/tasks/env/specs/themes/targets from other configs [optional]
 import:
   - ./some-dir/sake.yaml

 # Verify SSH host connections. Set this to true if you wish to circumvent verify host [optional]
 disable_verify_host: false

 # Set known_hosts_file path. Default is users ssh home directory [optional]
 # known_hosts_file: $HOME/.ssh/known_hosts

 # Set timeout for ssh connections in seconds
 # default_timeout: 20

 # Shell used for commands [optional]
 # If you use any other program than bash, zsh, sh, node, or python
 # then you have to provide the command flag if you want the command-line string evaluted
 # For instance: bash -c
 shell: bash

 # List of Servers
 servers:
   # Server name [required]
   media:
     # Server description [optional]
     desc: media server

     # Host [required]
     host: media.lan
     # one-line for setting user and port
     # host: samir@media.lan:22

     # Specify multiple hosts:
     # hosts:
     # - samir@192.168.0.1:22
     # - samir@l92.168.1.1:22

     # or use a host range generator
     # hosts: samir@192.168.[0:1].1:22

     # generate hosts by local command
     # inventory: echo samir@192.168.0.1:22 samir@192.168.1.1:22

     # Bastion [optional]
		 bastion: samir@192.168.1.1:2222

     # Bastions [optional]
     # bastions: [samir@192.168.1.1:2222, samir@192.168.1.2:3333]

     # User to connect as. It defaults to the current user [optional]
     user: samir

     # Port for ssh [optional]
     port: 22

     # Shell used for commands [optional]
     shell: bash

     # Run on localhost [optional]
     local: false

     # Set default working directory for task execution [optional]
     work_dir: ""

     # Set identity file. By default it will attempt to establish a connection using a SSH auth agent [optional]
     # sake respects users ssh config, so you can set auth credentials in the users ssh config
     identity_file: ./id_rsa

     # Set password. Accepts either a string or a shell command [optional]
     password: $(echo $MY_SECRET_PASSWORD)

     # List of tags [optional]
     tags: [remote]

     # List of server specific environment variables [optional]
     env:
       # Simple string value
       key: value

       # Shell command substitution (evaluated on localhost)
       date: $(date -u +"%Y-%m-%dT%H:%M:%S%Z")

 # List of environment variables that are available to all tasks
 env:
   # Simple string value
   AUTHOR: "alajmo"

   # Shell command substitution (evaluated on localhost)
   DATE: $(date -u +"%Y-%m-%dT%H:%M:%S%Z")

 # List of themes
 themes:
   # Theme name
   default:
     # Text options [optional]
     text:
       # Set host prefix for each line [optional]
       # Available variables: `.Name`, `.Index`, `.Host`, `.Port`, `.User`
       prefix: '{{ .Host }}'

       # Colors to alternate between for each server prefix [optional]
       # Available options: green, blue, red, yellow, magenta, cyan
       prefix_colors: ["green", "blue", "red", "yellow", "magenta", "cyan"]

       # Customize the task header that is printed before each task when output is set to text (to opt out, set it to empty string) [optional]
       # Available variables: `.Name`, `.Desc`, `.Index`, `.NumTasks`
       # Available methods: `.Style`, which takes in 1 or more parameters, first is the string to be styled, and the rest are styling options
       # Available styling options:
       #   Colors (prefix with `fg_` for foreground, and `bg_` for background): black, red, green, yellow, blue, magenta, cyan, white, hi_black, hi_red, hi_green, hi_yellow, hi_blue, hi_magenta, hi_cyan, hi_white
       #   Attributes: normal, bold, faint, italic, underline crossed_out
       header: '{{ .Style "TASK" "bold" }}{{ if ne .NumTasks 1 }} ({{ .Index }}/{{ .NumTasks }}){{end}}{{ if and .Name .Desc }} [{{.Style .Name "bold"}}: {{ .Desc }}] {{ else if .Name }} [{{ .Name }}] {{ else if .Desc }} [{{ .Desc }}] {{end}}'

       # Fill remaining spaces with a character after the header, if set to empty string, no filler characters will be displayed [optional]
       header_filler: "*"

     # Table options [optional]
     table:
       # Table style [optional]
       # Available options: ascii, connected-light
       style: ascii

       # Set host prefix [optional]
       # Available variables: `.Name`, `.Index`, `.Host`, `.Port`, `.User`
       prefix: '{{ .Host }}'

       # Border options for table output [optional]
       options:
         draw_border: false
         separate_columns: true
         separate_header: true
         separate_rows: false
         separate_footer: false

       # Color, attr, align, and format options [optional]
       # Available options for fg/bg: green, blue, red, yellow, magenta, cyan, hi_green, hi_blue, hi_red, hi_yellow, hi_magenta, hi_cyan
       # Available options for align: left, center, justify, right
       # Available options for attr: normal, bold, faint, italic, underline, crossed_out
       # Available options for format: default, lower, title, upper
       title:
         fg:
         bg:
         align:
         attr:
         format:

       header:
         fg:
         bg:
         align:
         attr:
         format:

       row:
         fg:
         bg:
         align:
         attr:
         format:

       footer:
         fg:
         bg:
         align:
         attr:
         format:

       border:
         header:
           fg:
           bg:
           attr:

         row:
           fg:
           bg:
           attr:

         row_alt:
           fg:
           bg:
           attr:

         footer:
           fg:
           bg:
           attr:

 # List of Specs [optional]
 specs:
   default:
     # Spec description
     desc: default spec

     # Print task description
     describe: false

     # Print list of hosts that will be targetted
     list_hosts: false

     # Order hosts [inventory|reverse_inventory|sorted|reverse_sorted|random]
     order: inventory

     # Omit showing loader when running tasks
     silent: false

     # Execution strategy [linear|host_pinned|free]
     strategy: linear

     # Number of hosts to run in parallel
     batch: 1

     # Number of hosts in percentage to run in parallel [0-100]
     # batch_p: 100

     # Max number of forks
     forks: 10000

     # Set task output [text|table|table-2|table-3|table-4|html|markdown|json|csv|none]
     output: text

     # Limit output [stdout|stderr|all]
     print: all

     # Hide task from auto-completion
     hidden: false

     # Continue task execution on errors
     ignore_errors: true

     # Stop task execution on any error
     any_errors_fatal: false

     # Max number of tasks to fail before aborting
     max_fail_percentage: 100

     # Ignore unreachable hosts
     ignore_unreachable: false

     # Omit empty rows for table output
     omit_empty_rows: false

     # Omit empty columns for table output
     omit_empty_columns: false

     # Show task reports [recap|rc|task|time|all]
     report: [recap]

     # Verbose turns on describe, list_hosts and report set to all
     verbose: false

     # Confirm invoked task before running
     confirm: false

     # Confirm each task before running
     step: false

 # List of targets [optional]
 targets:
   default:
     # Target description
     desc: ""

     # Target all hosts
     all: false

     # Specify hosts via server name
     servers: []

     # Specify hosts via server tags
     tags: []

     # Limit number of hosts to target
     limit: 0

     # Limit number of hosts to target in percentage
     limit_p: 100

     # Invert matching on hosts
     invert: false

     # Specify host regex
     regex: ""

 # List of tasks
 tasks:
   # Command ID [required]
   simple-1:
     # The name that will be displayed when executing or listing tasks. Defaults to task ID [optional]
     name: Simple

     # Script to run
     cmd: |
       echo "hello world"
     desc: simple command 1

   # Short-form for a command
   simple-2: echo "hello world"

   # Command ID [required]
   advanced-command:
     # The name that will be displayed when executing or listing tasks. Defaults to task ID [optional]
     name: Advanced Command

     # Task description [optional]
     desc: Advanced task

     # Specify theme [optional]
     theme: default

     # Spec reference [optional]
     # spec: default

     # Or specify specs inline
     spec:
       output: table
       ignore_errors: true
       ignore_unreachable: true
       any_errors_fatal: false
       omit_empty_rows: true
       omit_empty_columns: true

     # Target reference [optional]
     # target: default

     # Or specify targets inline
     target:
       all: true
       servers: [media]
       tags: [remote]
       limit: 1

     # List of environment variables [optional]
     env:
       # Simple string value
       release: v1.0.0

       # Shell command substitution
       num_lines: $(ls -1 | wc -l)

       # The following variables are available by default:
       #   S_NAME
       #   S_HOST
       #   S_USER
       #   S_PORT
       #   S_BASTION
       #   S_TAGS
       #   S_IDENTITY
       #   SAKE_DIR
       #   SAKE_PATH

     # Run on localhost [optional]
     local: false

     # Set default working directory for task [optional]
     work_dir: ""

     # Shell used for commands [optional]
     shell: bash

     # Each task can only define:
     # - a single cmd
     # - or a single task reference
     # - or a list of task references and commands

     # Single command
     cmd: |
       echo complex
       echo command

     # Task reference. work_dir and env variables are passed down
     task: simple-1

     # List of task references or commands
     tasks:
       # Command
       - name: inline-command
         cmd: echo "Hello World"
         ignore_errors: true
         work_dir: /tmp
         shell: bash
         env:
           foo: bar

       # Task reference. work_dir and env variables are passed down.
       # Nested task referencing is supported and will result in a
       # flat list of commands
       - task: simple-1
         ignore_errors: true
         work_dir: /tmp
         register: results
         env:
           foo: bar

       - name: output
         cmd: echo $results_stdout
.RE

.SH EXAMPLES

.TP
Initialize sake
.B ~ $ sake init

.nf
Initialized sake in /tmp
- Created sake.yaml

Following servers were added to sake.yaml

 Server    | Host
-----------+---------
 localhost | 0.0.0.0
.fi

.TP
List all servers
.B ~ $ sake list servers

.nf
 Server    | Host
-----------+---------
 localhost | 0.0.0.0
.fi

.TP
List all tasks
.B ~ $ sake list tasks

.nf
 Task | Description
------+-------------
 ping | Pong
.fi

.TP
Describe a task
.B ~ $ sake describe task ping

.nf
name: ping
desc: ping server
local: false
work_dir:
theme: default
target:
    all: true
spec:
    output: text
    ignore_unreachable: true
    omit_empty_rows: true
    omit_empty_columns: true
cmd:
    echo pong
.fi

.TP
Run a task for all servers with tag 'local'
.B ~ $ sake run many --tags local

.nf
TASK [ping: Pong] ********************

0.0.0.0 | pong
.fi

.TP
Run ad-hoc command for all servers
.B ~ $ sake exec --all --output table 'echo 123'

.nf
 Server    | Output
-----------+--------
 localhost | 123
.fi

.SH FILES

When running a command,
.B sake
will check the current directory and all parent directories for the following files: sake.yaml, sake.yml, .sake.yaml, .sake.yml.

Additionally, it will import (if found) a config file from:

.RS 2
.IP "\(bu" 2
Linux: \fB$XDG_CONFIG_HOME/sake/config.yaml\fR or \fB$HOME/.config/sake/config.yaml\fR if \fB$XDG_CONFIG_HOME\fR is not set.
.IP "\(bu" 2
Darwin: \fB$HOME/Library/Application/sake\fR
.RE

Both the config and user config can be specified via flags or environments variables.

.SH
ENVIRONMENT

.TP
.B SAKE_CONFIG
Override config file path

.TP
.B SAKE_USER_CONFIG
Override user config file path

.TP
.B SAKE_SSH_CONFIG
Override ssh config file path

.TP
.B SAKE_KNOWN_HOSTS_FILE
Override known_hosts file path

.TP
.B NO_COLOR
If this env variable is set (regardless of value) then all colors will be disabled

.SH BUGS

See GitHub Issues:
.UR https://github.com/alajmo/sake/issues
.ME .

.SH AUTHOR

.B sake
was written by Samir Alajmovic
.MT alajmovic.samir@gmail.com
.ME .
For updates and more information go to
.UR https://\:www.sakecli.com
sakecli.com
.UE .


docs/changelog.md
# Changelog

## 0.15.0

### Features

- Add support for multiple bastions

### Misc

- Update to go 1.20

### Fixes

- Fix issue where user/port was not set correctly when using shorthand format for host/user/port definition

## 0.14.0

### Features

- Add ability to modify prefix in text and table themes
- Hide tasks from auto-completion via spec attribute `hidden: true`
- Add print option to limit output to stdout|stderr
- Default to one of following identity files if no identity specified `~/.ssh/id_rsa`, `~/.ssh/id_ecdsa`, `~/.ssh/id_dsa`
- Add ability to modify default timeout for ssh connections

### Fixes

- [BREAKING CHANGE]: No more duplicate tasks, specs, targets, and themes
- Small fix when user config is specified but not found
- Fix some small validation issues with batch and batch-p
- A bunch of smaller fixes

## 0.13.0

### Features

- Add ability to register variables which are available to the next tasks
- Add option to ignore errors for indiviual tasks
- Add flag/spec `--list-hosts` option to list targetted hosts
- Support output options `csv`/`json`/`none`
- Add new task strategies: linear, host_pinned, free
  - `linear`: execute task for each host before proceeding to the next task (default)
  - `host_pinned`: executes tasks (serial) for a host before proceeding to the next host
  - `free`: tasks without waiting for other tasks
- Add host ordering
  - `inventory`: The order is as provided by the inventory
  - `reverse_inventory`: The order is the reverse of the inventory
  - `sorted`: Hosts are alphabetically sorted by host
  - `reverse_sorted`: Hosts are sorted by host in reverse alphabetical order
  - `random`: Hosts are randomly ordered
- Determine number of hosts to run in parallel
  - `batch`: specify number of hosts
  - `batch_p`: specify number of hosts in percentage
  - `forks`: max number of concurrent processes
- Add option to display reports at end of tasks by using `--report` flag or specifying it in `spec` definition
  - `recap`: show basic report
  - `rc`: show return code for each host and task
  - `task`: show task status for each host and task
  - `time`: show time report for each host and task
  - `all`: show all reports
- Add confirm/step task capability
  - `confirm`: for the root task
  - `step`: per task and host

### Fixes

- Fix omitting attribute `align` when creating a theme
- Abort tasks prematurely when running in parallel and AnyErrorsFatal set to true
- Fix server range (previously `[2:100]` didn't work as strings were compared)
- Fix empty error for non existing working directory and update how work_dir works

### Minor

- Add option to omit empty columns via flag `--omit-empty-columns` and spec `omit_empty_columns`
- Add option to specify target and spec via flags `--target`/`--spec`
- Add description to targets and specs
- Add server identity to environment variables
- Add silent/describe attribute to spec definition
- Add ssh user flag option

### Changes

- [**BREAKING CHANGE**]: Deprecated the parallel flag, use batch/batch_p/forks instead
- [**BREAKING CHANGE**]: Rename default environment variables from `SAKE_SERVER_*` to `S_*`, and remove task default environment variables
- [**BREAKING CHANGE**]: Shorthand flag for silent is now `Q`
- Switch to default shell when evaluating inventory
- If no command name is set on nested tasks, assign `task-$i` instead of `task`
- If `--limit` flag is higher than available hosts, then select all hosts filtered
- Update flag sorting
- Rename `--omit-empty` to `--omit-empty-rows`
- Building `sake` with go 1.19

## 0.12.1

### Fixes

- Fix port out of range when using shorthand format for hosts

## 0.12.0

### Features

- Add hosts keyword that supports having multiple hosts per server definition
  - Specify as a list
  - Specify as a string containing range (`192.168.0.[1:10:2]`)
  - Use `inventory` attribute (`kubectl get nodes`)
- Add silent flag to supress `Running...` spinner when running tasks
- Support connection string instead of 3 fields: `user@host:port`
- Support resolving IdentityFile in ssh config (`~/.ssh/config`)
- Support resolving Includes in ssh config (`~/.ssh/config`)
- Support glob pattern for Hosts (`Host *`)
- Add bastion headers to list servers
- Add flags/target config `--limit` & `--limit-p` to limit number of servers task is run on
- Add filtering servers on host regex
- Add invert flag on filtering servers
- Add flag `--all-headers` for tasks and servers
- Add sub-commands edit/list/describe [specs|targets]
- Add 3 new table outputs (table-2, table-3, table-4)
- [BREAKING CHANGE]: Simplified theme config, now it only accepts manipulation of rows and headers, not specific properties

### Fixes

- Use IdentitiesOnly if user specifies a IdentityFile
- Default to `Name`, if description is not set, in auto-completion for tasks
- Support lowercase ssh config keys (previously they had to be PascalCase)

### Deprecated

- [BREAKING CHANGE]: Removed environment variables `SAKE_IDENTITY_FILE` and `SAKE_PASSWORD`, users can use flags instead

## 0.11.0

### Fixes

- Fix not being able to parse ssh config if match keyword found [35](https://github.com/alajmo/sake/pull/35)

### Features

- Support Bastion/jump host [32](https://github.com/alajmo/sake/pull/32)

## 0.10.3

### Fixes

- Previously known_hosts didn't work correctly when specifying port other than 22
- Fix authentication failures [32](https://github.com/alajmo/sake/pull/30)

### Features

- Resolve hosts from ssh_config
- Support hashing known_host entries

## 0.10.2

### Fixes

- Allow duplicate hosts
- Fix correct exit code on remote/local task errors (#27)
- Fix local WorkDir when it's not explicitly set

## 0.10.1

### Fixes

- Small fix for WorkDir being related to calling file when server is local

## 0.10.0

### Fixes

- Fix issue where ipv6 was not added correctly to known_hosts (brackets without ip)
- Fix TTY in sub-tasks
- Only task or cmd allowed in inline `tasks` definition

### Changes


core/dao/config.go
package dao

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"text/template"

	"github.com/jedib0t/go-pretty/v6/text"
	"gopkg.in/yaml.v3"

	"github.com/alajmo/sake/core"
)

var (
	ACCEPTABLE_FILE_NAMES = []string{"sake.yaml", "sake.yml", ".sake.yaml", ".sake.yml"}

	DEFAULT_SHELL = "bash -c"

	DEFAULT_TIMEOUT = uint(20)

	DEFAULT_THEME = Theme{
		Name:  "default",
		Table: DefaultTable,
		Text:  DefaultText,
	}

	DEFAULT_TARGET = Target{
		Name:    "default",
		All:     false,
		Servers: []string{},
		Tags:    []string{},
		Regex:   "",
		Invert:  false,
		Limit:   0,
		LimitP:  0,
	}

	DEFAULT_SPEC = Spec{
		Name:              "default",
		Desc:              "the default spec",
		Describe:          false,
		ListHosts:         false,
		Order:             "inventory",
		Silent:            false,
		Hidden:            false,
		Strategy:          "linear",
		Batch:             0,
		BatchP:            0,
		Forks:             10000,
		Output:            "text",
		MaxFailPercentage: 0,
		AnyErrorsFatal:    true,
		IgnoreErrors:      false,
		IgnoreUnreachable: false,
		OmitEmptyRows:     false,
		OmitEmptyColumns:  false,
		Report:            []string{"recap"},
		Verbose:           false,
		Confirm:           false,
		Step:              false,
		Print:             "all",
	}
)

type Config struct {
	SSHConfigFile     *string
	DefaultTimeout    uint
	DisableVerifyHost bool
	KnownHostsFile    string
	Shell             string
	Envs              []string
	Themes            []Theme
	Specs             []Spec
	Targets           []Target
	Servers           []Server
	Tasks             []Task
	Path              string
}

type ConfigYAML struct {
	// Internal
	Path           string  `yaml:"-"`
	Dir            string  `yaml:"-"`
	UserConfigFile *string `yaml:"-"`

	// Intermediate
	DisableVerifyHost *bool     `yaml:"disable_verify_host"`
	DefaultTimeout    *uint     `yaml:"default_timeout"`
	KnownHostsFile    *string   `yaml:"known_hosts_file"`
	Shell             string    `yaml:"shell"`
	Import            yaml.Node `yaml:"import"`
	Env               yaml.Node `yaml:"env"`
	Themes            yaml.Node `yaml:"themes"`
	Specs             yaml.Node `yaml:"specs"`
	Targets           yaml.Node `yaml:"targets"`
	Servers           yaml.Node `yaml:"servers"`
	Tasks             yaml.Node `yaml:"tasks"`

	contextLine int `yaml:"-"`
}

func (c *ConfigYAML) GetContext() string {
	return c.Path
}

func (c *ConfigYAML) GetContextLine() int {
	return c.contextLine
}

// Function to read sake configs.
func ReadConfig(configFilepath string, userConfigPath string, sshConfigFile string, noColor bool) (Config, error) {
	CheckUserNoColor(noColor)
	var configPath string

	userConfigFile, err := getUserConfigFile(userConfigPath)
	if err != nil {
		return Config{}, err
	}

	sshConfigPath, err := getSSHConfigPath(sshConfigFile)
	if err != nil {
		return Config{}, err
	}

	// Try to find config file in current directory and all parents
	if configFilepath != "" {
		filename, err := filepath.Abs(configFilepath)
		if err != nil {
			return Config{}, err
		}

		configPath = filename
	} else {
		wd, err := os.Getwd()
		if err != nil {
			return Config{}, err
		}

		// Check first cwd and all parent directories, then if not found,
		// check if env variable SAKE_CONFIG is set, and if not found
		// return no config found
		filename, err := core.FindFileInParentDirs(wd, ACCEPTABLE_FILE_NAMES)
		if err != nil {
			val, present := os.LookupEnv("SAKE_CONFIG")
			if present {
				filename = val
			} else {
				return Config{}, err
			}
		}

		filename, err = filepath.Abs(filename)
		if err != nil {
			return Config{}, err
		}

		configPath = filename
	}

	dat, err := os.ReadFile(configPath)
	if err != nil {
		return Config{}, err
	}

	// Found config, now try to read it
	var configYAML ConfigYAML

	configYAML.Path = configPath
	configYAML.Dir = filepath.Dir(configPath)
	configYAML.UserConfigFile = userConfigFile

	err = yaml.Unmarshal(dat, &configYAML)
	if err != nil {
		re := ResourceErrors[ConfigYAML]{Resource: &configYAML, Errors: []error{err}}
		return Config{}, FormatErrors(re.Resource, re.Errors)
	}

	config, configErr := configYAML.parseConfig()
	config.SSHConfigFile = sshConfigPath
	config.CheckConfigNoColor()

	if configErr != nil {
		return config, configErr
	}

	return config, nil
}

func getUserConfigFile(userConfigPath string) (*string, error) {
	// Flag
	if userConfigPath != "" {
		if _, err := os.Stat(userConfigPath); err != nil {
			return nil, fmt.Errorf("user config not found: %w", err)
		}
		return &userConfigPath, nil
	}

	// Env
	val, present := os.LookupEnv("SAKE_USER_CONFIG")
	if present {
		if _, err := os.Stat(val); err != nil {
			return nil, fmt.Errorf("user config not found: %w", err)
		}
		return &val, nil
	}

	// Default
	defaultUserConfigDir, _ := os.UserConfigDir()

	defaultUserConfigPath := filepath.Join(defaultUserConfigDir, "sake", "config.yaml")
	if _, err := os.Stat(defaultUserConfigPath); err == nil {
		return &defaultUserConfigPath, nil
	}

	defaultUserConfigPath = filepath.Join(defaultUserConfigDir, "sake", "config.yml")
	if _, err := os.Stat(defaultUserConfigPath); err == nil {
		return &defaultUserConfigPath, nil
	}

	return nil, nil
}

func getSSHConfigPath(sshConfigPath string) (*string, error) {
	// Flag
	if sshConfigPath != "" {
		if _, err := os.Stat(sshConfigPath); err == nil {
			return &sshConfigPath, nil
		} else {
			return &sshConfigPath, err
		}
	}

	// Env
	val, present := os.LookupEnv("SAKE_SSH_CONFIG")
	if present {
		return &val, nil
	}

	// User SSH config
	if home, err := os.UserHomeDir(); err == nil {
		userSSHConfigFile := filepath.Join(home, ".ssh", "config")
		if _, err := os.Stat(userSSHConfigFile); err == nil {
			return &userSSHConfigFile, nil
		}
	}

	// Global SSH config
	globalSSHConfig := "/etc/ssh/ssh_config"
	if _, err := os.Stat(globalSSHConfig); err == nil {
		return &globalSSHConfig, nil
	}

	return nil, nil
}

// Open sake config in editor
func (c *Config) EditConfig() error {
	return openEditor(c.Path, -1)
}

func openEditor(path string, lineNr int) error {
	editor, found := os.LookupEnv("EDITOR")
	if !found {
		return &core.NoEditorEnv{}
	}

	var args []string

	if lineNr > 0 {
		switch editor {
		case "vim":
			args = []string{fmt.Sprintf("+%v", lineNr), path}
		case "vi":
			args = []string{fmt.Sprintf("+%v", lineNr), path}
		case "emacs":
			args = []string{fmt.Sprintf("+%v", lineNr), path}
		case "nano":
			args = []string{fmt.Sprintf("+%v", lineNr), path}
		case "code": // visual studio code
			args = []string{"--goto", fmt.Sprintf("%s:%v", path, lineNr)}
		case "idea": // Intellij
			args = []string{"--line", fmt.Sprintf("%v", lineNr), path}
		case "subl": // Sublime
			args = []string{fmt.Sprintf("%s:%v", path, lineNr)}
		case "atom":
			args = []string{fmt.Sprintf("%s:%v", path, lineNr)}
		case "notepad-plus-plus":
			args = []string{"-n", fmt.Sprintf("%v", lineNr), path}
		default:
			args = []string{path}
		}
	} else {
		args = []string{path}
	}

	editorBin, err := exec.LookPath(editor)
	if err != nil {
		return err
	}

	cmd := exec.Command(editorBin, args...)
	cmd.Env = os.Environ()
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	err = cmd.Run()
	if err != nil {
		return err
	}

	return nil
}

// Open sake config in editor and optionally go to line matching the task name
func (c *Config) EditTask(name string) error {
	configPath := c.Path
	if name != "" {
		task, err := c.GetTask(name)
		if err != nil {
			return err
		}
		configPath = task.context
	}

	dat, err := os.ReadFile(configPath)
	if err != nil {
		return err
	}

	type ConfigTmp struct {
		Tasks yaml.Node
	}

	var configTmp ConfigTmp
	err = yaml.Unmarshal(dat, &configTmp)
	if err != nil {
		return err
	}

	lineNr := 0
	if name == "" {
		lineNr = configTmp.Tasks.Line - 1
	} else {
		for _, task := range configTmp.Tasks.Content {
			if task.Value == name {
				lineNr = task.Line
				break
			}
		}
	}

	return openEditor(configPath, lineNr)
}

// Open sake config in editor and optionally go to line matching the server name
func (c *Config) EditServer(name string) error {
	var group string
	configPath := c.Path
	if name != "" {
		server, err := c.GetServerByGroup(name)
		if err != nil {
			return err
		}
		group = server.Group
		configPath = server.context
	}

	dat, err := os.ReadFile(configPath)
	if err != nil {
		return err
	}

	type ConfigTmp struct {
		Servers yaml.Node
	}

	var configTmp ConfigTmp
	err = yaml.Unmarshal(dat, &configTmp)
	if err != nil {
		return err
	}

	lineNr := 0
	if name == "" {
		lineNr = configTmp.Servers.Line - 1
	} else {
		for _, server := range configTmp.Servers.Content {
			if server.Value == group {
				lineNr = server.Line
				break
			}
		}
	}

	return openEditor(configPath, lineNr)
}

// Open sake config in editor and optionally go to line matching the target name
func (c *Config) EditTarget(name string) error {
	configPath := c.Path
	if name != "" {
		target, err := c.GetTarget(name)
		if err != nil {
			return err
		}
		configPath = target.context
	}

	dat, err := os.ReadFile(configPath)
	if err != nil {
		return err
	}

	type ConfigTmp struct {
		Targets yaml.Node
	}

	var configTmp ConfigTmp
	err = yaml.Unmarshal(dat, &configTmp)
	if err != nil {
		return err
	}

	lineNr := 0
	if name == "" {
		lineNr = configTmp.Targets.Line - 1
	} else {
		for _, target := range configTmp.Targets.Content {
			if target.Value == name {
				lineNr = target.Line
				break
			}
		}
	}

	return openEditor(configPath, lineNr)
}

// Open sake config in editor and optionally go to line matching the spec name
func (c *Config) EditSpec(name string) error {
	configPath := c.Path
	if name != "" {
		spec, err := c.GetSpec(name)
		if err != nil {
			return err
		}
		configPath = spec.context
	}

	dat, err := os.ReadFile(configPath)
	if err != nil {
		return err
	}

	type ConfigTmp struct {
		Specs yaml.Node
	}

	var configTmp ConfigTmp
	err = yaml.Unmarshal(dat, &configTmp)
	if err != nil {
		return err
	}

	lineNr := 0
	if name == "" {
		lineNr = configTmp.Specs.Line - 1
	} else {
		for _, spec := range configTmp.Specs.Content {
			if spec.Value == name {
				lineNr = spec.Line
				break
			}
		}
	}

	return openEditor(configPath, lineNr)
}

func InitSake(args []string) ([]Server, error) {
	// Choose to initialize sake in a different directory
	// 1. absolute or
	// 2. relative or
	// 3. working directory
	var configDir string
	if len(args) > 0 && filepath.IsAbs(args[0]) {
		// absolute path
		configDir = args[0]
	} else if len(args) > 0 {
		// relative path
		wd, err := os.Getwd()
		if err != nil {
			return []Server{}, err
		}
		configDir = filepath.Join(wd, args[0])
	} else {
		// working directory
		wd, err := os.Getwd()
		if err != nil {
			return []Server{}, err
		}
		configDir = wd
	}

	err := os.MkdirAll(configDir, os.ModePerm)
	if err != nil {
		return []Server{}, err
	}

	configPath := filepath.Join(configDir, "sake.yaml")
	if _, err := os.Stat(configPath); err == nil {
		return []Server{}, &core.AlreadySakeDirectory{Dir: configDir}
	}

	rootName := "localhost"
	rootHost := "0.0.0.0"
	rootServer := Server{Name: rootName, Host: rootHost}
	servers := []Server{rootServer}
	funcMap := template.FuncMap{
		"serverItem": func(name string, host string) string {
			var txt = name + ":"
			txt = txt + "\n    host: " + host
			txt = txt + "\n    local: true"
			return txt
		},
	}

	tmpl, err := template.New("init").Funcs(funcMap).Parse(`servers:
  {{- range .}}
  {{ (serverItem .Name .Host) }}
  {{ end }}
tasks:
  ping:
    desc: Pong
    cmd: echo "pong"
`,
	)
	if err != nil {
		return []Server{}, err
	}

	// Create sake.yaml
	f, err := os.Create(configPath)
	if err != nil {
		return []Server{}, err
	}

	err = tmpl.Execute(f, servers)
	if err != nil {
		return []Server{}, err
	}

	f.Close()

	fmt.Println("\nInitialized sake in", configDir)
	fmt.Println("- Created sake.yaml")

	return servers, nil
}

func (c *Config) ParseInventory(userArgs []string) error {
	var servers []Server
	var shell = DEFAULT_SHELL
	if c.Shell != "" {
		shell = c.Shell
	}
	shell = core.FormatShell(shell)

	for _, s := range c.Servers {
		if s.Inventory != "" {
			hosts, err := core.EvaluateInventory(shell, s.context, s.Inventory, s.Envs, userArgs)
			if err != nil {
				return err
			}

			if len(hosts) == 0 {
				return fmt.Errorf("inventory %s returned 0 hosts", s.Name)
			}

			for i, host := range hosts {
				server, err := CreateInventoryServers(host, i, s, userArgs)
				if err != nil {
					return err
				}

				servers = append(servers, server)
			}

		} else {
			servers = append(servers, s)
		}
	}

	c.Servers = servers

	return nil
}

func CheckUserNoColor(noColorFlag bool) {
	_, present := os.LookupEnv("NO_COLOR")
	if noColorFlag || present {
		text.DisableColors()
	}
}

func (c *Config) CheckConfigNoColor() {
	for _, env := range c.Envs {
		name := strings.Split(env, "=")[0]
		if name == "NO_COLOR" {
			text.DisableColors()
		}
	}
}
